node/worker/worker.go

```diff
@@ -258,16 +276,51 @@ func (w *Worker) IncomingReceipts() []*types.CXReceiptsProof {
 	return w.current.incxs
 }
 
-// CommitWithCrossLinks generate a new block with cross links for the new txs.
-func (w *Worker) CommitWithCrossLinks(sig []byte, signers []byte, viewID uint64, coinbase common.Address, crossLinks []byte) (*types.Block, error) {
+// ProposeShardStateWithoutBeaconSync proposes the next shard state for next epoch.
+func (w *Worker) ProposeShardStateWithoutBeaconSync() types.ShardState {
+	if !core.ShardingSchedule.IsLastBlock(w.current.header.Number.Uint64()) {
+		return nil
+	}
+	nextEpoch := new(big.Int).Add(w.current.header.Epoch, common.Big1)
+	return core.GetShardState(nextEpoch)
+}
+
+// FinalizeNewBlock generate a new block for the next consensus round.
+func (w *Worker) FinalizeNewBlock(sig []byte, signers []byte, viewID uint64, coinbase common.Address, crossLinks types.CrossLinks, shardState types.ShardState) (*types.Block, error) {
 	if len(sig) > 0 && len(signers) > 0 {
 		copy(w.current.header.LastCommitSignature[:], sig[:])
 		w.current.header.LastCommitBitmap = append(signers[:0:0], signers...)
 	}
 	w.current.header.Coinbase = coinbase
 	w.current.header.ViewID = new(big.Int)
 	w.current.header.ViewID.SetUint64(viewID)
-	w.current.header.CrossLinks = crossLinks
+
+	// Cross Links
+	if crossLinks != nil && len(crossLinks) != 0 {
```

Comment:
OT but FWIW: Can we just define the `CrossLinks` header field not as `[]byte` but as `types.CrossLink` so that we do not have to do manual serialization/deserialization?  Or do we need to differentiate the “`CrossLinks` field absent” and “`CrossLinks` field present but is an empty slice” case?  (For `ShardState` that was the case from what I remember… we may need to revisit that as well.)

Reply:
I think the serialization/deserailziation should be handled separately, rather than using the general serialization of the header. Just to have more flexibility and better control on the header data.

Reply:
I see.  Flexibility and control may be better achieved by implementing `rlp.Encoder` and `rlp.Decoder` interface IMO but this PR is not the right place to do that anyway.